### ISSUE TITLE
Doctrine DBAL 4 multiple changes and deprecations

### ISSUE THREAD

**Issue by Shotman** [RamyHakam/multi_tenancy_bundle#35]
I can't create migrations for the tenants and it seems like DBAL deprecated/changed a few things.

- The namespace `Doctrine\DBAL\Events` doesn't exist anymore.
- In `TenantConnection.php` the `_driver` attribute is now just `driver`
- In the same file `_eventManager` doesn't exist anymore and has to be instanciated with `new EventManager()`

I'll keep digging and add information on this issue if I find anything useful

Here is my composer info :

```
beganovich/snappdf                  v4.0.3           Convert webpages or HTML into the PDF file using Chromium or Google Chrome.
behat/transliterator                v1.5.0           String transliterator
brick/math                          0.11.0           Arbitrary-precision arithmetic library
chrisguitarguy/request-id-bundle    v6.0.0           Add request IDs to to your Symfony requests.
composer/semver                     3.4.0            Semver library that offers utilities, version constraint parsing and validation.
defuse/php-encryption               v2.4.0           Secure PHP Encryption Library
doctrine/annotations                2.0.1            Docblock Annotations Parser
doctrine/cache                      2.2.0            PHP Doctrine Cache library is a popular cache implementation that supports many different drivers such as...   
doctrine/collections                2.2.1            PHP Doctrine Collections library that adds additional functionality on top of PHP arrays.
doctrine/common                     3.4.3            PHP Doctrine Common project is a library that provides additional functionality that other Doctrine proje...   
doctrine/dbal                       4.0.1            Powerful PHP database abstraction layer (DBAL) with many features for database schema introspection and m...   
doctrine/deprecations               1.1.3            A small layer on top of trigger_error(E_USER_DEPRECATED) or PSR-3 logging with options to disable all dep...   
doctrine/doctrine-bundle            2.12.0           Symfony DoctrineBundle
doctrine/doctrine-migrations-bundle 3.3.0            Symfony DoctrineMigrationsBundle
doctrine/event-manager              2.0.0            The Doctrine Event Manager is a simple PHP event system that was built to be used with the various Doctri...
doctrine/inflector                  2.0.10           PHP Doctrine Inflector is a small library that can perform string manipulations with regard to upper/lowe...   
doctrine/instantiator               2.0.0            A small, lightweight utility to instantiate objects in PHP without invoking their constructors
doctrine/lexer                      3.0.1            PHP Doctrine Lexer parser library that can be used in Top-Down, Recursive Descent Parsers.
doctrine/migrations                 3.7.4            PHP Doctrine Migrations project offer additional functionality on top of the database abstraction layer (...   
doctrine/orm                        3.0.0            Object-Relational-Mapper for PHP
doctrine/persistence                3.3.2            The Doctrine Persistence project is a set of shared interfaces and functionality that the different Doctr...   
doctrine/sql-formatter              1.2.0            a PHP SQL highlighting library
egulias/email-validator             4.0.2            A library for validating emails against several RFCs
fdekker/log-viewer-bundle           v0.7.2           A log viewer bundle for Symfony 6 and up
gedmo/doctrine-extensions           v3.14.0          Doctrine behavioral extensions
hakam/multi-tenancy-bundle          v2.6.0           Symfony bundle to extend doctrine to support db switcher and multi tenants
masterminds/html5                   2.9.0            An HTML5 parser and serializer.
mediamonks/doctrine-extensions      3.1.1            Doctrine2 behavioral extensions which allows to transform (encrypt, decrypt, hash) your data automatically     
monolog/monolog                     3.5.0            Sends your logs to files, sockets, inboxes, databases and various web services
myclabs/deep-copy                   1.11.1           Create deep copies (clones) of your objects
nikic/php-parser                    v5.0.2           A PHP parser written in PHP
paragonie/random_compat             v9.99.100        PHP 5.x polyfill for random_bytes() and random_int() from PHP 7
phar-io/manifest                    2.0.4            Component for reading phar.io manifest information from a PHP Archive (PHAR)
phar-io/version                     3.2.1            Library for handling version information and constraints
phpdocumentor/reflection-common     2.2.0            Common reflection classes used by phpdocumentor to reflect the code structure
phpdocumentor/reflection-docblock   5.4.0            With this component, a library can provide support for annotations via DocBlocks or otherwise retrieve in...   
phpdocumentor/type-resolver         1.8.2            A PSR-5 based resolver of Class names, Types and Structural Element Names
phpstan/phpdoc-parser               1.28.0           PHPDoc parser with support for nullable, intersection and generic types
phpunit/php-code-coverage           9.2.31           Library that provides collection, processing, and rendering functionality for PHP code coverage information.   
phpunit/php-file-iterator           3.0.6            FilterIterator implementation that filters files based on a list of suffixes.
phpunit/php-invoker                 3.1.1            Invoke callables with a timeout
phpunit/php-text-template           2.0.4            Simple template engine.
phpunit/php-timer                   5.0.3            Utility class for timing
phpunit/phpunit                     9.6.19           The PHP Unit Testing framework.
psr/cache                           3.0.0            Common interface for caching libraries
psr/clock                           1.0.0            Common interface for reading the clock.
psr/container                       2.0.2            Common Container Interface (PHP FIG PSR-11)
psr/event-dispatcher                1.0.0            Standard interfaces for event handling.
psr/link                            2.0.1            Common interfaces for HTTP links
psr/log                             3.0.0            Common interface for logging libraries
ramsey/collection                   2.0.0            A PHP library for representing and manipulating collections.
ramsey/uuid                         4.7.5            A PHP library for generating and working with universally unique identifiers (UUIDs).
seaworn/symfony-htmx-bundle         dev-main 0447e65 Htmx integration for Symfony
sebastian/cli-parser                1.0.2            Library for parsing CLI options
sebastian/code-unit                 1.0.8            Collection of value objects that represent the PHP code units
sebastian/code-unit-reverse-lookup  2.0.3            Looks up which function or method a line of code belongs to
sebastian/comparator                4.0.8            Provides the functionality to compare PHP values for equality
sebastian/complexity                2.0.3            Library for calculating the complexity of PHP code units
sebastian/diff                      4.0.6            Diff implementation
sebastian/environment               5.1.5            Provides functionality to handle HHVM/PHP environments
sebastian/exporter                  4.0.6            Provides the functionality to export PHP variables for visualization
sebastian/global-state              5.0.7            Snapshotting of global state
sebastian/lines-of-code             1.0.4            Library for counting the lines of code in PHP source code
sebastian/object-enumerator         4.0.4            Traverses array structures and object graphs to enumerate all referenced objects
sebastian/object-reflector          2.0.4            Allows reflection of object attributes, including inherited and non-public ones
sebastian/recursion-context         4.0.5            Provides functionality to recursively process PHP variables
sebastian/resource-operations       3.0.4            Provides a list of PHP built-in functions that operate on resources
sebastian/type                      3.2.1            Collection of value objects that represent the types of the PHP type system
sebastian/version                   3.0.2            Library that helps with managing the version number of Git-hosted PHP projects
stof/doctrine-extensions-bundle     v1.10.1          Integration of the gedmo/doctrine-extensions with Symfony
symfony/asset                       v6.4.3           Manages URL generation and versioning of web assets such as CSS stylesheets, JavaScript files and image f...
symfony/asset-mapper                v6.4.5           Maps directories of assets & makes them available in a public directory with versioned filenames.
symfony/browser-kit                 v6.4.3           Simulates the behavior of a web browser, allowing you to make requests, click on links and submit forms p...   
symfony/cache                       v6.4.6           Provides extended PSR-6, PSR-16 (and tags) implementations
symfony/cache-contracts             v3.4.2           Generic abstractions related to caching
symfony/clock                       v6.4.5           Decouples applications from the system clock
symfony/config                      v6.4.6           Helps you find, load, combine, autofill and validate configuration values of any kind
symfony/console                     v6.4.6           Eases the creation of beautiful and testable command line interfaces
symfony/css-selector                v6.4.3           Converts CSS selectors to XPath expressions
symfony/debug-bundle                v6.4.3           Provides a tight integration of the Symfony VarDumper component and the ServerLogCommand from MonologBrid...   
symfony/dependency-injection        v6.4.6           Allows you to standardize and centralize the way objects are constructed in your application
symfony/deprecation-contracts       v3.4.0           A generic function and convention to trigger deprecation notices
symfony/doctrine-bridge             v6.4.6           Provides integration for Doctrine with various Symfony components
symfony/doctrine-messenger          v6.4.6           Symfony Doctrine Messenger Bridge
symfony/dom-crawler                 v6.4.4           Eases DOM navigation for HTML and XML documents
symfony/dotenv                      v6.4.4           Registers environment variables from a .env file
symfony/error-handler               v6.4.6           Provides tools to manage errors and ease debugging PHP code
symfony/event-dispatcher            v6.4.3           Provides tools that allow your application components to communicate with each other by dispatching event...   
symfony/event-dispatcher-contracts  v3.4.2           Generic abstractions related to dispatching event
symfony/expression-language         v6.4.3           Provides an engine that can compile and evaluate expressions
symfony/filesystem                  v6.4.6           Provides basic utilities for the filesystem
symfony/finder                      v6.4.0           Finds files and directories via an intuitive fluent interface
symfony/flex                        v2.4.5           Composer plugin for Symfony
symfony/form                        v6.4.6           Allows to easily create, process and reuse HTML forms
symfony/framework-bundle            v6.4.6           Provides a tight integration between Symfony components and the Symfony full-stack framework
symfony/http-client                 v6.4.6           Provides powerful methods to fetch HTTP resources synchronously or asynchronously
symfony/http-client-contracts       v3.4.2           Generic abstractions related to HTTP clients
symfony/http-foundation             v6.4.4           Defines an object-oriented layer for the HTTP specification
symfony/http-kernel                 v6.4.6           Provides a structured process for converting a Request into a Response
symfony/intl                        v6.4.3           Provides access to the localization data of the ICU library
symfony/mailer                      v6.4.6           Helps sending emails
symfony/maker-bundle                v1.58.0          Symfony Maker helps you create empty commands, controllers, form classes, tests and more so you can forge...   
symfony/messenger                   v6.4.6           Helps applications send and receive messages to/from other applications or via message queues
symfony/mime                        v6.4.6           Allows manipulating MIME messages
symfony/monolog-bridge              v6.4.4           Provides integration for Monolog with various Symfony components
symfony/monolog-bundle              v3.10.0          Symfony MonologBundle
symfony/notifier                    v6.4.3           Sends notifications via one or more channels (email, SMS, ...)
symfony/options-resolver            v6.4.0           Provides an improved replacement for the array_replace PHP function
symfony/orm-pack                    v2.4.0           A pack for the Doctrine ORM
symfony/password-hasher             v6.4.4           Provides password hashing utilities
symfony/phpunit-bridge              v7.0.6           Provides utilities for PHPUnit, especially user deprecation notices management
symfony/polyfill-intl-grapheme      v1.29.0          Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu           v1.29.0          Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn           v1.29.0          Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer    v1.29.0          Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring           v1.29.0          Symfony polyfill for the Mbstring extension
symfony/polyfill-php83              v1.29.0          Symfony polyfill backporting some PHP 8.3+ features to lower PHP versions
symfony/process                     v6.4.4           Executes commands in sub-processes
symfony/property-access             v6.4.6           Provides functions to read and write from/to an object or array using a simple string notation
symfony/property-info               v6.4.6           Extracts information about PHP class' properties using metadata of popular sources
symfony/routing                     v6.4.6           Maps an HTTP request to a set of configuration variables
symfony/runtime                     v6.4.3           Enables decoupling PHP applications from global state
symfony/security-bundle             v6.4.6           Provides a tight integration of the Security component into the Symfony full-stack framework
symfony/security-core               v6.4.3           Symfony Security Component - Core Library
symfony/security-csrf               v6.4.3           Symfony Security Component - CSRF Library
symfony/security-http               v6.4.4           Symfony Security Component - HTTP Integration
symfony/serializer                  v6.4.6           Handles serializing and deserializing data structures, including object graphs, into array structures or ...   
symfony/service-contracts           v3.4.2           Generic abstractions related to writing services
symfony/stimulus-bundle             v2.16.0          Integration with your Symfony app & Stimulus!
symfony/stopwatch                   v6.4.3           Provides a way to profile code
symfony/string                      v6.4.4           Provides an object-oriented API to strings and deals with bytes, UTF-8 code points and grapheme clusters ...   
symfony/translation                 v6.4.4           Provides tools to internationalize your application
symfony/translation-contracts       v3.4.2           Generic abstractions related to translation
symfony/twig-bridge                 v6.4.6           Provides integration for Twig with various Symfony components
symfony/twig-bundle                 v6.4.4           Provides a tight integration of Twig into the Symfony full-stack framework
symfony/ux-turbo                    v2.16.0          Hotwire Turbo integration for Symfony
symfony/validator                   v6.4.6           Provides tools to validate values
symfony/var-dumper                  v6.4.6           Provides mechanisms for walking through any arbitrary PHP variable
symfony/var-exporter                v6.4.6           Allows exporting any serializable PHP data structure to plain PHP code
symfony/web-link                    v6.4.3           Manages links between resources
symfony/web-profiler-bundle         v6.4.4           Provides a development tool that gives detailed information about the execution of any request
symfony/yaml                        v6.4.3           Loads and dumps YAML files
theseer/tokenizer                   1.2.3            A small library for converting tokenized PHP source code into XML and potentially other formats
twig/extra-bundle                   v3.8.0           A Symfony bundle for extra Twig extensions
twig/twig                           v3.8.0           Twig, the flexible, fast, and secure template language for PHP
webmozart/assert                    1.11.0           Assertions to validate method input/output with nice error messages.
```

**Comment by Shotman** [comment]
According to this : https://[raw.githubusercontent.com/doctrine/dbal/c70bae9e69e0e543044d5e282a624d9b3cbea6fb/UPGRADE.md](https://raw.githubusercontent.com/doctrine/dbal/c70bae9e69e0e543044d5e282a624d9b3cbea6fb/UPGRADE.md)
the postConnect event is deprecated and now a custom middleware must be used

**Comment by Shotman** [comment]
As a temporary fix forcing `doctrine/dbal` to version `^3.6` and running `composer update -W` 
But everything breaks with bdal 4.0

**Comment by RamyHakam** [comment]
thanks For rasing and updating  this issue , I will create a PR for that
